[IP-254] Receipts for payments

mihaimatei opened the pull request:
I've implemented receipts tied to payment method, as follows:
1. Created a new invoicing/numbering group for receipts
2. Added the option to choose the numbering group for a payment method.
3. Made payment method data available to the invoice PDF template and if a numbering group is assigned to used payment method, it will display a receipt block.
   It is designed to support partial payments, so multiple receipts for each invoice.
   **I've added changes to the PDF template from branch 1.4.6, and marked the added lines between <!-- IP-254 --> comment :|, sorry for that.**

Database changes summary:

```
ALTER TABLE `ip_payment_methods` ADD COLUMN `receipt_group_id` INT(11) NULL;
ALTER TABLE `ip_payments` ADD COLUMN `receipt_number` VARCHAR(100) NULL;
```

**The SQL queries should be moved into the proper next release sql setup file, now it's in 017_1.4.4-nextRelease.sql**

More info in the Jira ticket comment: https://development.invoiceplane.com/browse/IP-254

Pitma committed:
Corrected Fileissue if language contains special characters (i.e. greek)

nerdoc committed:
fix logo upload image width/height proportions

No matter how the proportions of the logo were, it was always displayed as fixed 80x80px - this looked distorted if the logo is e.g. 200x40px.

lommes committed:
Added invoice footer also to quotes

tbartholomes committed:
IP-310 - Taxes and products do not respect separator and decimal

tbartholomes committed:
IP-312 - Current logged in user not displayed

lommes committed:
removed superfluous quote

Kovah committed:
Merge pull request #250 from lommes/IP-312

IP-312 - Current logged in user not displayed

lommes committed:
IP-315 - Array definition incompatible with PHP < 5.4

Kovah committed:
Merge pull request #251 from lommes/IP-315

IP-315 - Array definition incompatible with PHP < 5.4

lommes committed:
IP-291 - Let CodeIgniter also accept new domain names

Kovah committed:
Merge pull request #252 from lommes/IP-291

IP-291 - Let CodeIgniter also accept new domain names

Kovah committed:
Merge pull request #249 from lommes/IP-310

IP-310 - Taxes and products do not respect separator and decimal

Kovah committed:
Merge pull request #248 from lommes/master

Added invoice footer also to quotes

Kovah committed:
Merge pull request #247 from nerdoc/patch-1

fix logo upload image width/height proportions

Commit:
IP-316 Vagrant and ansible should be allowed for dev work
Getting the start of a vm for invoiceplane

Commit:
IP-316 Vagrant and ansible should be allowed for dev work
Updating readme

Kovah committed:
Merge pull request #255 from cgsmith/feature/Vagrant-ansible

IP-316 Feature/vagrant ansible

Kovah committed:
Added sql file for updates

lommes committed:
IP-293 - Custom title for public view of quotes / invoices

Commit:
IP-317 Integrate additional payment gateways
The start of the payment gateway integrations.  Work in progress!

Kovah committed:
Merge pull request #256 from lommes/IP-293

IP-293 - Custom title for public view of quotes / invoices

nerdoc committed:
Merge pull request #2 from InvoicePlane/master

udpate to upstream

nerdoc committed:
show real server name at end of setup

nerdoc committed:
Merge pull request #4 from nerdoc/nerdoc-patch-servername

show real server name at end of setup

Commit:
show http(s) protocol at setup complete hint

Kovah committed:
Merge pull request #257 from nerdoc/master

show (http(s)) protocol in setup complete procedure

Commit:
IP-318

Commit:
IP-318

Forgot to add support for item_name

Commit:
IP-318

Kovah committed:
Merge pull request #261 from aalwash/development

IP-318 - Support for variables in product / item names

pauluz committed:
correct label name for "Values with taxes" in "Sales by Date" report

Kovah committed:
Merge pull request #262 from pauluz/develop

correct label name for "Values with taxes" in "Sales by Date" report

pauluz committed:
Add id="content" to main DIV in project view/index - because Options menu was cut by the layout (in FF)

nerdoc committed:
Add (disabled) option to force HTTPS access

It's much more convenient if you just need to remove some comment hashs.

Kovah committed:
Merge pull request #264 from nerdoc/patch-2

Add (disabled) option to force HTTPS access

Commit:
remove database.php from git, create it during setup

Commit:
Link to /setup within warning message if database.php not exists.

Kovah committed:
Merge pull request #266 from nerdoc/without-database-php

remove database.php from git, create it during setup

pauluz committed:
Merge remote-tracking branch 'upstream/development' into develop

Commit:
IPT-45 - Invoice now contains page numbers by default

Commit:
IPT-45 Fixed tab to spaces

Kovah committed:
Merge pull request #267 from aalwash/development

IPT-45 - Add page numbers to template

pauluz committed:
Merge remote-tracking branch 'upstream/development' into develop

pauluz committed:
files in IP for running tests

pauluz committed:
changes in core files needed to run PHPUnit

pauluz committed:
tests for number_helper.php

pauluz committed:
updated README.md

pauluz committed:
add new line to better look

Kovah committed:
Merge pull request #263 from pauluz/develop

Add id="content" to main DIV in project view/index - because Options menu was cut by the layout (in FF)

Commit:
IP-320 - Add page numbering for customer footer (easy to disable, by
modifing assets/default/css/custom.css)

Commit:
IP-321 - Fixed syntax error

nerdoc committed:
typo: added comma.

Kovah committed:
Merge pull request #269 from nerdoc/patch-1

typo: added comma.

pauluz committed:
correct id="content" in recurring invoice layout

nerdoc committed:
fixed table id/class mix error.

Kovah committed:
Merge pull request #271 from nerdoc/patch-2

fixed table id/class mix error.

Kovah committed:
Merge pull request #270 from pauluz/IP-314-recurring-invoice

correct id="content" in recurring invoice layout

Commit:
adds company name to username in header

targets IP-312 - better fix

Commit:
change invoice/quote heder widths 70/30 to 60/40

Translated labels now have enough space.

Kovah committed:
Merge pull request #272 from nerdoc/user-email

adds company name to username in header

Kovah committed:
Merge pull request #273 from nerdoc/fix-invoice-widths

change invoice/quote heder widths 70/30 to 60/40

pauluz committed:
Fix array syntax error in upload module

Kovah committed:
Merge pull request #274 from pauluz/IP-upload-array-error

Fix array syntax error in upload module

pauluz committed:
hide custom fields if empty in clients module (view)

Kovah committed:
Merge pull request #275 from pauluz/IP-hide-custom-fields-if-empty-in-clients

Hide custom fields if empty in clients module (view)

Commit:
Added custom field types functionality to InvoicePlane custom fields

Commit:
Pulled unsynced changes

Commit:
Removed headnotes

Kovah committed:
Merge pull request #276 from visualcookie/development

IP-46 - Added custom field types functionality to InvoicePlane custom fields

Commit:
Merge remote-tracking branch 'upstream/development' into development

Commit:
IP-320 Moved CSS from css/custom.css to sass/templates.scss

irimia committed:
Dashboard "This/Last Quarter" listing fix

This/Last Quarter from current year.

Commit:
Fix: There is now translation entry with the key 'quote_number'. Change to fit.

Commit:
Also check if  mpdf/ttfontdata, mpdf/graph_cache are writable. See mpdf documenation.

Kovah committed:
Merge pull request #280 from aalwash/development

IP-320 - Add page numbering for customer footer

irimia committed:
row() missing

Kovah committed:
Merge pull request #279 from unknownsoftware/master

Dashboard "This/Last Quarter" listing fix

Kovah committed:
Merge pull request #282 from Silver-Connection/master

Fix: IP-329 and IP-330

pauluz committed:
fix set_setting()

Kovah committed:
Merge pull request #283 from pauluz/fix-settings

fix set_setting()

Commit:
IP-46: Removed printed array

Commit:
IP-46: Removed duplicated code

Kovah committed:
Merge pull request #285 from visualcookie/development

IP-46: Removed duplicated code and the printed out array

pauluz committed:
use jquery-1.11.2 in sessions module (not 1.11.1)

pauluz committed:
correct setting value for pdf template

Kovah committed:
Merge pull request #286 from pauluz/fix-jquery-in-sessions

use jquery-1.11.2 in sessions module (not 1.11.1)

Kovah committed:
Merge pull request #287 from pauluz/fix-template-settings-in-guest

correct setting value for pdf template in guest module

Commit:
Update email input type and remember values

I've added type="email" to show the email keyboard on mobile devices. Also made it remember what you submit so that if you submit a wrong password, you don't have to re-write your email address and vice-versa.

Kovah committed:
Merge pull request #289 from WebNinjrDesign/patch-2

Update email input type and remember values

Commit:
Standardized HTML5 Doctype

Standardized the HTML5 doctype declarations across the app.
Lower case was chosen to stick to the standards in HTML5 boilerplate.

HTML5 Boilerplate: https://github.com/h5bp/html5-boilerplate
Doctype Discussion:
https://github.com/h5bp/html5-boilerplate/issues/1522

Commit:
Updated 404 Page

An updated 404 error page. Cleaner, simpler with no container.
- Now responsive + viewport set.
- Removed ‘404’ from the page title. No need for this, users just need
to know that the page was not found. The number is a side note.
- Added ‘InvoicePlane’ to the title, so that the user knows its apart
of the InvoicePlane environment.
- Updated the ‘code’ element styling to be a bit simpler. I wasn’t sure
if this was actually needed, as I only see ‘p’ tags coming through.

Kovah committed:
Merge pull request #290 from WebNinjrDesign/doctype-update

Standardized HTML5 Doctype

Kovah committed:
Merge pull request #291 from WebNinjrDesign/update-404

Updated 404 Page

Commit:
Update error_404.php

- Removed 'style' tags type 'attribute' as text/css is default in HTML5 and unneeded.
- Set the page title to the $header variable, as this is probably best for future translations etc.

Commit:
Removed error code

Removed the error code from the $heading variable as the user has no
interest in the error code, only what has happened i.e. “Page Not
Found”.

Commit:
Updated database error page

Similar styling to 404 error page except:
1. Line length is much longer to accommodate for error messages.
2. <p> tags have more spacing for readability of several messages.

Also, code tag styling has been removed, as this doesn’t seem needed on
this page. Please correct me if I am wrong.

Commit:
General updates to error pages.

Kovah committed:
Merge pull request #292 from WebNinjrDesign/update-error-pages

Update error pages

Commit:
Reformatted, Standardized Quotes

- Removed excess spacing before array items.
- Standardized quotes. Previously mix of single and double quotes, now
just single quotes.

Commit:
Fixed spacing

Fixed spacing between values

Commit:
Fixed tabs

Replaced tabs with spaces to fix spacing again.

Commit:
Removed closing php tag

- Removed un-needed closing php tags from bottom of files.

Commit:
Standardized quotes

Standardized all string quotes to single quotes.

Commit:
Align values

Aligned array values for readability and easier editing.

Commit:
Removed my comments

Just removing my change comments in the file

Commit:
Removed closing php tag

Commit:
- Removed script tag type attribute. Default is text/javascript in HTML5
- Standardised single quotes in the JS

Commit:
- Changed if to more fronted code friendly  if() : endif;
- Updated ‘b’ tag to ‘strong’. Better semantics.

Commit:
- removed script type attribute. default in HTML5 is text/javascript.
- Standardized single quotes in JS

Commit:
Standardised single quotes

Commit:
Tidied array spacing and layout for readability.

Commit:
Fix menu bug

- Fixed an issue where a stray div was left after removing code (PR:
#285) and was causing the menu in the view clients page to not function.

- Removed stray quote

Commit:
Standardised quotes

Commit:
Aligned values

Commit:
Fixed spacing

Commit:
Removed trailing comma

Commit:
Alignment and readability

Moved items with array values to their own array declaration. Easier to
read and maintain.
Made SVG a single string, not an array.

Commit:
Standarized quotes

Commit:
Fixed spacing

Commit:
Fixed spacing

Commit:
Update PHP error page

Commit:
Reformatted HTML

Commit:
Standardised quotes + Spacing

Commit:
replace endif with endforeach

Commit:
- Fixed double quote issue
- adjusted spacing
- removed b tags in favour of th tags for table headers.
- fixed self closing br and hr tags.
- made conditionals and loops if() : endif; format. Better for
front-end readability

Commit:
- Updated spacing and used alternative foreach () : endforeach; syntax.

Commit:
Used alternative if() : endif; syntax

Commit:
- Removed script tag type attribute. Defaults to text/javascript in
html5.
- alternative syntax for if and foreach statements.
- Tidied switch formatting

Commit:
Fixed indenting and use of alternative foreach syntax

Commit:
Fixed array spacing

Kovah committed:
Merge pull request #294 from WebNinjrDesign/fix-menu-bug

Fixed menu bug

Kovah committed:
Merge pull request #295 from WebNinjrDesign/config-tidy

Config files update

Kovah committed:
Merge pull request #296 from WebNinjrDesign/error-page-update

Update PHP error page

flaiw committed:
IP-162 continued: add task to invoice + changing task status when task is added to invoice / removed from invoice or when invoice is deleted.

Commit:
Moved email pdf attachment setting out of SMTP tab and made it a global
email setting. The setting already applies to all email methods, its
just that it was tucked away in the SMTP tab.

Commit:
Added tooltips to sidebar items

Was not obvious what the sidebar items were, they now have tooltips.

flaiw committed:
Added tax rate to tasks
Do not allow changing client of an invoice when invoice refers to tasks
Do not allow task editing when referred to by an invoice
Message when user tries to delete task with status invoiced

Kovah committed:
Merge pull request #303 from WebNinjrDesign/sidebar-tooltips

Added tooltips to sidebar items

Kovah committed:
Merge pull request #302 from WebNinjrDesign/email-pdf-setting

Fix: Email PDF attachement now global

Kovah committed:
Merge branch 'IP-162-continued' of https://github.com/flaiw/InvoicePlane into flaiw-IP-162-continued, solved merge conflicts

Kovah committed:
Code reformatting according to PSR1/PSR2 and moved language strings to correct location

Kovah committed:
Update to Codeigniter 2.2.4

dodomorandi committed:
Added support for decimals for price and amounts

Kovah committed:
Merge pull request #301 from dodomorandi/decimal_price_amount

Added support for decimals for price and amounts

aalwash committed:
Fixed syntax error -- See switch warning and about spaces -- http://php.net/manual/en/control-structures.alternative-syntax.php

Kovah committed:
Merge pull request #305 from aalwash/development

Fixed syntax error in clients form

cgsmith committed:
Adding composer files for omnipay integration

cgsmith committed:
Adding composer file

Kovah committed:
Moved development part out of the README into CONTRIBUTING

Kovah committed:
Merge branch 'feature/stripe-integration' of https://github.com/cgsmith/InvoicePlane into cgsmith-feature/stripe-integration, solved merge conflicts

Kovah committed:
Merge branch 'cgsmith-feature/stripe-integration' into development

Kovah committed:
Code corrections because of broken merge regarding the Omnipay integration

jaskoola committed:
Invoice attachment improvment: delete confirmation and download button

Kovah committed:
Merge pull request #308 from jaskoola/development

Invoice attachment improvement: delete confirmation and download button

Kovah committed:
IP-317 - Removed PayPal and Stripe as they are already included in omnipay/omnipay

Kovah committed:
Merge pull request #309 from Kovah/development

IP-317 - Removed PayPal and Stripe from composer.json

jaskoola committed:
Invoice attachment improvement: replace whitespaces to '_' in filename after upload in download and delete buttons, fix formating and remove alert window in download button

Kovah committed:
Merge pull request #310 from jaskoola/development

Invoice attachment improvement: replace whitespaces to '_' in filename

jaskoola committed:
IP-348 - Improve email templates to support EmailToSMS services

Kovah committed:
IP-317 - Refactored the merchant naming to online payments and extended the payment handling, @todo finish payment process

Kovah committed:
IP-317 - Refactored the merchant naming to online payments and extended the payment handling, @todo finish payment process

Kovah committed:
IP-317 - Refactored the merchant naming to online payments and extended the payment handling, @todo finish payment process

Kovah committed:
IP-317 - Refactored the merchant naming to online payments and extended the payment handling, @todo finish payment process

Kovah committed:
Extended the contribution documentation

Kovah committed:
Merge pull request #312 from Kovah/development

Optimized the readme and contribution guide

Kovah committed:
IP-346 - Fixes payment form handling that caused problems with pre-set values

Kovah committed:
IP-346 - Fixes payment form handling that caused problems with pre-set values

Kovah committed:
Merge branch 'development' of github.com:Kovah/InvoicePlane into development

Kovah committed:
Merge pull request #313 from Kovah/development

IP-346 - Fixes payment form handling that caused problems with pre-set values

Kovah committed:
IP-328 - Removed format_thousands_separator() as it prevents using the space character as a thousand separator

Kovah committed:
Merge pull request #314 from Kovah/development

IP-328 - Removed format_thousands_separator() function

Kovah committed:
IP-331 - Recurring invoices now process CSS and HTML in email templates

Kovah committed:
Merge pull request #315 from Kovah/development

IP-331 - Recurring invoices now process CSS and HTML in email templates

Kovah committed:
Fixed label definitions in recurring invoices

Kovah committed:
Merge pull request #316 from Kovah/development

Fixed label definitions in recurring invoices

Kovah committed:
Merge branch 'development' into IP-317

Kovah committed:
Updated the README with new badges

Kovah committed:
Merge pull request #317 from Kovah/development

Updated the README with new badges

Kovah committed:
Merge branch 'IP-317' of github.com:Kovah/InvoicePlane into IP-317

Kovah committed:
Merge pull request #311 from jaskoola/development

Improve email templates to support EmailToSMS services

Rayvw committed:
IP-360 - Implemented mPDF template file

Kovah committed:
Merge branch 'IP-317' of github.com:Kovah/InvoicePlane into IP-317

Kovah committed:
Implemented support for onsite payments with credit card (tested with Stripe), offsite payments are still missing

Kovah committed:
Implemented support for onsite payments with credit card (tested with Stripe), offsite payments are still missing

Kovah committed:
Merge branch 'IP-317' of github.com:Kovah/InvoicePlane into IP-317

Kovah committed:
Merge pull request #320 from Kovah/IP-317

IP-317 (Online Payment overhaul)

jaskoola committed:
IP-354 -  Recurring Invoice: set "Invoices Due After" and "Email Template" in recuring invoices

Kovah committed:
Merge pull request #321 from jaskoola/development

IP-354 -  Recurring Invoice: set "Invoices Due After" and "Email Template" in recuring invoices

jaskoola committed:
IP-353: EmailToSMS feature is missing SQL table modifications, which is causing errors on save

jaskoola committed:
IP-354 - database update

Kovah committed:
Merge pull request #322 from jaskoola/development

IP-353: EmailToSMS feature is missing SQL table modifications, which is causing errors on save

Rayvw committed:
IP-360 - Added upload field to settings>invoices for PDF template

Kovah committed:
Merge pull request #318 from Rayvw/development

IP-360 - Implemented mPDF template file

Rayvw committed:
IP-104 - Custom fields are now copied when invoice or quote is duplicated

Kovah committed:
Merge pull request #328 from Rayvw/feature/ip-104

IP-104 - Custom fields are not copied when Invoice is duplicated

profago committed:
Update mdl_import.php

With this small add the client's import process fills all the fields.

if you accept this change you may have to update also the columns reference in https://wiki.invoiceplane.com/en/1.0/system/importing-data adding the relative two columns.

Kovah committed:
Merge pull request #330 from profago/patch-1

Update mdl_import.php

dantefff committed:
Added suport for having a custom template for each invoice group

Conflicts:
	application/helpers/pdf_helper.php

Kovah committed:
Merge pull request #334 from dantefff/development

IP-109: Added suport for having a custom template for each invoice group

rahajiki committed:
New lines to html br in guest quote view

New lines to html br in guest quote view

rahajiki committed:
Semicolon after parentheses

Kovah committed:
Merge pull request #345 from rahajiki/development

New lines to html br in guest quote view

mihaimatei committed:
IP-254 Receipts feature for payments

christoferd committed:
IP-400 fix null password PDF protection

 IP-400 fix null password PDF protection

Kovah committed:
Merge pull request #350 from christoferd/patch-1

 IP-400 fix null password PDF protection

Kovah commented:
Could you please do a [rebase](https://help.github.com/articles/about-git-rebase/) of your branch to prevent duplicate commits from other users.

Kovah commented:
@mihaimatei

mihaimatei commented:
@Kovah , sorry for the delay, I've rebase first time got me with a lot of conflicts and others commits, I don't know why, but probably because I've started based on v1.4.4 or something was not up to date.
I've rebased it again, checked the changes, squashed my commits and now looks fine.

kalimerre commented on application/views/invoice_templates/pdf/InvoicePlane.php:
```diff
@@ -0,0 +1,282 @@
+                    <?php echo format_currency($invoice_tax_rate->invoice_tax_rate_amount); ?>
+                </td>
+            </tr>
+        <?php endforeach ?>
```
It miss the ";" after the endforeach

kalimerre commented on application/views/invoice_templates/pdf/InvoicePlane.php:
```diff
@@ -0,0 +1,282 @@
+            <h1 class="invoice-title"><?php echo lang('receipt') . ' ' . $payment->receipt_number; ?></h1>
+            <div id="company">
+                <div><b><?php echo $invoice->user_company; ?></b></div>
+                <?php if ($invoice->user_vat_id)
```
Would be great to use the same format as line 19 for example. It's a template and most of users are not developers so it will be hard to understand the condition for them
```
if(condition) {
   echo "blabla";
}
```

UnderDogg replied:
Needs label : 1.6 or 1.5.11 or something. Needs Milestone?

UnderDogg commented:
Needs label : 1.6 or 1.5.11 or something. Needs Milestone?

nielsdrost7 commented:
superseeded by https://github.com/InvoicePlane/InvoicePlane/pull/1523/
